### PR TITLE
[Feature] get compensation voltages from FAIMS experiment

### PR DESCRIPTION
**Pull request by enetz**
# Description

PR to OpenMS/split_by_FAIMS_CV branch, not to develop.

Initial commits for PeakMap splitting. Simple function to get compensation voltages and test file for splitting.

# Checklist:
- [x] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes. (Tick if no updates were necessary.)

# How can I get additional information on failed tests during CI:
If your PR is failing you can check out 
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. If you click in the column that lists the failed tests you will get detailed error messages.

### CODE REVIEW

File: src/openms/source/IONMOBILITY/MSRunIMSplitter.cpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    splitPeakMap.resize(CVs.size());
+
+    for (PeakMap::Iterator it = exp.begin(); it != exp.end(); ++it)
+    {
+      //TODO copy MetaData to new PeakMaps? or keep original as reference for merging back later?
+      for (Size i = 0; i < CVs.size(); i++)
+      {
+        if (it->getDriftTime() == CVs[i])
+        {
+          splitPeakMap[i].addSpectrum(*it);
```
Comment (timosachsenberg) [inline src/openms/source/IONMOBILITY/MSRunIMSplitter.cpp:59]: split_map[i].emplace_back(std::move(*it));

File: src/openms/source/IONMOBILITY/MSRunIMSplitter.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+// $Authors: Eugen Netz $
+// --------------------------------------------------------------------------
+
+#include <OpenMS/IONMOBILITY/MSRunIMSplitter.h>
+#include <OpenMS/IONMOBILITY/FAIMSHelper.h>
+#include <OpenMS/KERNEL/MSExperiment.h>
+
+namespace OpenMS
+{
+  std::vector<PeakMap> MSRunIMSplitter::splitByFAIMSCV(PeakMap& exp)
```
Comment (timosachsenberg) [inline src/openms/source/IONMOBILITY/MSRunIMSplitter.cpp:41]: PeakMap&&

Comment (timosachsenberg) [inline src/openms/source/IONMOBILITY/MSRunIMSplitter.cpp:41]: // splits exp (and clears) them

File: src/openms/include/OpenMS/IONMOBILITY/FAIMSHelper.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    {
+    public:
+      virtual ~FAIMSHelper() {}
+
+      /**
+        @brief Get all FAIMS compensation voltages that occur in a PeakMap
+
+        @param exp The PeakMap with FAIMS data
+      */
+      static std::set<double> getCompensationVoltages(PeakMap& exp);
```
Comment (cbielow) [inline src/openms/include/OpenMS/IONMOBILITY/FAIMSHelper.h:60]: `const PeakMap&`

File: src/openms/include/OpenMS/IONMOBILITY/MSRunIMSplitter.h [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    public:
+      virtual ~MSRunIMSplitter() {}
+
+      /**
+        @brief Splits a PeakMap into one PeakMap per FAIMS CV value
+
+        This only works with a PeakMap that has a FAIMS compensation voltage
+        associated with each spectrum.
+        The spectra from the original PeakMap are moved to new PeakMaps,
+        so the orinǵinal PeakMap is unusable afterwards.
```
Comment (cbielow) [inline src/openms/include/OpenMS/IONMOBILITY/MSRunIMSplitter.h:59]: typo here

File: src/openms/include/OpenMS/IONMOBILITY/MSRunIMSplitter.h [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+        @param exp The PeakMap
+      */
```
Comment (cbielow) [inline src/openms/include/OpenMS/IONMOBILITY/MSRunIMSplitter.h:62]: this function should IMHO throw an Exception if there is no FAIMS data available.

Maybe add a function: `bool isFAIMSExperiment(const PeakMap& exp)` which users can call beforehand to make sure its safe.

File: src/openms/source/IONMOBILITY/FAIMSHelper.cpp [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+#include <algorithm>
+
+namespace OpenMS
+{
+  std::set<double> FAIMSHelper::getCompensationVoltages(PeakMap& exp)
+  {
+    std::set<double> CVs;
+
+    // is this FAIMS data?
+    if (exp.getSpectra()[0].getDriftTimeUnit() == MSSpectrum::DriftTimeUnit::FAIMS_COMPENSATION_VOLTAGE)
```
Comment (cbielow) [inline src/openms/source/IONMOBILITY/FAIMSHelper.cpp:46]: segfaults if map is empty.

Comment (cbielow) [reply src/openms/source/IONMOBILITY/FAIMSHelper.cpp:46]: add `if (!exp.empty() && ...`

File: src/openms/source/IONMOBILITY/FAIMSHelper.cpp [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+// $Authors: Eugen Netz $
+// --------------------------------------------------------------------------
+
+#include <OpenMS/IONMOBILITY/FAIMSHelper.h>
+#include <OpenMS/KERNEL/MSExperiment.h>
+#include <algorithm>
+
+namespace OpenMS
+{
+  std::set<double> FAIMSHelper::getCompensationVoltages(PeakMap& exp)
```
Comment (cbielow) [inline src/openms/source/IONMOBILITY/FAIMSHelper.cpp:41]: maybe add to docs that the result is empty if this is not a FAIMS experiment.
On the other hand, maybe throwing an exception would be even safer...

Comment (timosachsenberg) [reply src/openms/source/IONMOBILITY/FAIMSHelper.cpp:41]: I think documenting empty would be fine. Does what is says :)

**Comment by cbielow** [comment]
any news here?

I'd like to build on top of this PR soonish :)

Also, we could merge into develop, right?

**Comment by timosachsenberg** [comment]
if you promise to add your suggestions ;) we can merge

**Comment by cbielow** [comment]
superceded by https://github.com/OpenMS/OpenMS/pull/5302

**Pull request closed without merging**